src/systems/filecoin_mining/storage_mining/storage_mining_subsystem.go

```diff
@@ -12,58 +12,114 @@ import (
+	params[1] = addr.Serialize_Address(workerAddr)
+	params[2] = libp2p.Serialize_PeerID(peerId)
+
+	sysActor, ok := state.GetActor(addr.SystemActorAddr)
+	Assert(ok)
+
+	unsignedCreationMessage := &msg.UnsignedMessage_I{
+		From_:       ownerAddr,
+		To_:         addr.StoragePowerActorAddr,
+		Method_:     ai.Method_StoragePowerActor_CreateStorageMiner,
+		Params_:     params,
+		CallSeqNum_: sysActor.CallSeqNum(),
+		Value_:      pledgeAmt,
+		GasPrice_:   0,
+		GasLimit_:   msg.GasAmount_SentinelUnlimited(),
+	}
+
+	var workerKey filcrypto.SigKeyPair // sms._keyStore().WorkerKey()
+	signedMessage, err := msg.Sign(unsignedCreationMessage, workerKey)
+	if err != nil {
+		return nil, err
+	}
+
+	err = sms.FilecoinNode().SubmitMessage(signedMessage)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: WAIT for block reception
+	util.IMPL_TODO()
+
+	var storageMinerAddr addr.Address
+	// and set in key store appropriately
+	return storageMinerAddr, nil
 }
 
 func (sms *StorageMiningSubsystem_I) HandleStorageDeal(deal deal.StorageDeal) {
 	sms.SectorIndex().AddNewDeal(deal)
 	// stagedDealResponse := sms.SectorIndex().AddNewDeal(deal)
 	// TODO: way within a node to notify different components
-	// markeet.StorageProvider().NotifyStorageDealStaged(&storage_provider.StorageDealStagedNotification_I{
+	// market.StorageProvider().NotifyStorageDealStaged(&storage_provider.StorageDealStagedNotification_I{
 	// 	Deal_:     deal,
 	// 	SectorID_: stagedDealResponse.SectorID(),
 	// })
 }
 
-func (sms *StorageMiningSubsystem_I) _generateOwnerAddress(workerPubKey filcrypto.PublicKey) addr.Address {
-	panic("TODO")
-}
-
 func (sms *StorageMiningSubsystem_I) CommitSectorError() deal.StorageDeal {
 	panic("TODO")
 }
 
 // triggered by new block reception and tipset assembly
 func (sms *StorageMiningSubsystem_I) OnNewBestChain() {
-	sms._tryLeaderElection()
+	sms._runMiningCycle()
 }
 
 // triggered by wall clock
 func (sms *StorageMiningSubsystem_I) OnNewRound() {
-	sms._tryLeaderElection()
+	sms._runMiningCycle()
+}
+
+func (sms *StorageMiningSubsystem_I) _runMiningCycle() {
+	chainHead := sms._blockchain().BestChain().HeadTipset()
+	sma := sms._getStorageMinerActorState(chainHead.StateTree(), sms._keyStore().MinerAddress())
+
+	if sma.ChallengeStatus().CanBeElected(chainHead.Epoch() + 1) {
+		ePoSt := sms._tryLeaderElection(chainHead.StateTree(), sma)
+		if ePoSt != nil {
+			// Randomness for ticket generation in block production
+			randomness1 := sms._consensus().GetTicketProductionRand(sms._blockchain().BestChain(), sms._blockchain().LatestEpoch())
+			newTicket := sms.PrepareNewTicket(randomness1, sms._keyStore().MinerAddress())
+
+			sms._blockProducer().GenerateBlock(ePoSt, newTicket, chainHead, sms._keyStore().MinerAddress())
+		}
+	} else if sma.ChallengeStatus().IsChallenged() {
+		sPoSt := sms._trySurprisePoSt(chainHead.StateTree(), sma)
+		// TODO: how to set these?
+		var gasLimit msg.GasAmount
+		var gasPrice = actor.TokenAmount(0)
```

Comment:
@jzimmerman @zixuanzh assuming some sort of IO or set param for the above

Reply:
Let's leave this as `IMPL_TODO` for now also -- the concrete parameters are dependent on how we ultimately decide to configure SurprisePoSt gas accounting, but the structure is pretty clear I think.